docs: rewrite parts of readme with more clarity (#161)
AEGEE/MyAEGEE: README.md

## Pre-requisites: URL mapping

You are encouraged to edit the `/etc/hosts` file on the HOST machine (on windows: `C:\Windows\system32\drivers\etc\hosts`) to add the entry:

> Vagrant case: `192.168.168.168 appserver.test my.appserver.test traefik.appserver.test portainer.appserver.test`

Reviewer: ```suggestion
Vagrant case: `192.168.168.168 appserver.test my.appserver.test traefik.appserver.test portainer.appserver.test pgadmin.appserver.test`
```
Let's put it here too, it won't hurt. Same below for the pure docker case

Author: I left it out intentionally 'as an exercise' for the person to do it themselves, as you see below pgadmin is 'the' example i picked to add a new host, domain, and service. What do you think?

Reviewer: I must say it's not the place I'm looking for exercises. Not in the setup guide. Others can weigh on this too.

Reviewer 2: EDIT: I was not awake yet. I would add pgadmin here as well. This way the most important subdomains are in the hosts file already

# Under the hood

> `Virtualbox` is a utility that lets people creating virtual machines (VM) on your computer. The created VM is the GUEST. The computer you run Virtualbox on is the HOST

Reviewer: I really like this sentence and all the GUEST/HOST additions. I think it should come way earlier in the file though, maybe after the /etc/hosts file edition. That way we understand better the sentence 
>Note: if you use Vagrant, Docker will be already automatically on the virtual machine.

which makes total sense now. This way, we can also explain that if we want to see the services, either we use the portainer UI, but if we're familiar with docker commands, we should expect to see the containers in the GUEST machine, that we can access using `vagrant ssh`. If we want to use docker commands, we must run them on the GUEST.

I think it could be nice to have an intermediate step to check that things are working as expected, before the big step in the end "open your browser on the HOST". This step could be:
```
vagrant ssh
docker ps
```
should yield an output like
```
username@computername:~/Documents/aegee/MyAEGEE$ vagrant ssh
Welcome to Ubuntu 18.04.4 LTS (GNU/Linux 4.15.0-111-generic x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

 System information disabled due to load higher than 1.0

 * Kubernetes 1.19 is out! Get it in one command with:

     sudo snap install microk8s --channel=1.19 --classic

   https://microk8s.io/ has docs and details.

97 packages can be updated.
70 updates are security updates.

This system is built by the Bento project by Chef Software
More information can be found at https://github.com/chef/bento
Last login: Fri Sep 18 14:13:43 2020 from 10.0.2.2
 appserver[/vagrant]$ [master !?] docker ps
CONTAINER ID        IMAGE                                  COMMAND                  CREATED             STATUS                 PORTS                                      NAMES
cbe689ea9ee2        aegee/discounts:dev                    "docker-entrypoint.s…"   5 hours ago         Up 5 hours (healthy)   8084/tcp                                   myaegee_discounts_1
ed1c20a4e65a        aegee/statutory:dev                    "docker-entrypoint.s…"   5 hours ago         Up 5 hours (healthy)   8084/tcp                                   myaegee_statutory_1
f60476fa4a6c        aegee/events:dev                       "docker-entrypoint.s…"   5 hours ago         Up 5 hours (healthy)   8084/tcp                                   myaegee_events_1
06add303acd0        aegee/core:dev                         "docker-entrypoint.s…"   5 hours ago         Up 5 hours (healthy)   8084/tcp                                   myaegee_core_1
72a7cb2d4100        postgres:10                            "docker-entrypoint.s…"   5 hours ago         Up 5 hours             5432/tcp                                   myaegee_postgres-statutory_1
316ced849d6f        postgres:10                            "docker-entrypoint.s…"   5 hours ago         Up 5 hours             5432/tcp                                   myaegee_postgres-core_1
f3110b11d9c6        postgres:10                            "docker-entrypoint.s…"   5 hours ago         Up 5 hours             5432/tcp                                   myaegee_postgres-discounts_1
9985f6cd3042        postgres:10                            "docker-entrypoint.s…"   5 hours ago         Up 5 hours             5432/tcp                                   myaegee_postgres-events_1
623356e7ef0d        dpage/pgadmin4:4.23                    "/entrypoint.sh"         6 hours ago         Up 6 hours             80/tcp, 443/tcp, 5050/tcp                  myaegee_pgadmin_1
0728da470558        erikdubbelboer/phpredisadmin:v1.11.4   "tini -- php -S 0.0.…"   6 hours ago         Up 6 hours             80/tcp                                     myaegee_redisadmin_1
8263f843df15        portainer/portainer:1.22.1             "/portainer -H unix:…"   6 hours ago         Up 6 hours             9000/tcp                                   myaegee_portainer_1
d056ce2c92e3        swaggerapi/swagger-ui:v3.28.0          "/docker-entrypoint.…"   6 hours ago         Up 6 hours             80/tcp, 8080/tcp                           myaegee_swagger_1
4ee964dede44        traefik:v1.7.4-alpine                  "/entrypoint.sh --we…"   6 hours ago         Up 6 hours (healthy)   0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp   myaegee_traefik_1
30963e2faf63        aegee/portal:dev                       "docker-php-entrypoi…"   6 hours ago         Up 6 hours             80/tcp                                     myaegee_portal_1
49485e19dc87        aegee/frontend:dev                     "nginx"                  6 hours ago         Up 6 hours             80/tcp, 8083/tcp                           myaegee_frontend_1
e81bcc38bf76        aegee/nginx-static:latest              "nginx -g 'daemon of…"   6 hours ago         Up 6 hours (healthy)   80/tcp                                     myaegee_statutory-static_1
dc2ea7b33c97        aegee/nginx-static:latest              "nginx -g 'daemon of…"   6 hours ago         Up 6 hours (healthy)   80/tcp                                     myaegee_events-static_1

```
what do we see here? That we can connect to vagrant, and that a bunch of containers are up. I think it's nice to introduce the `vagrant ssh` here already, because we'll see later that we should use `make start` on the GUEST.

Author: > intermediate step to check that things are working as expected

well, it's not intermediate, because I suggest to see if the website is up by looking at the website itself :)

However you convinced me with the rest of the sentence, i.e. that we are introducing vagrant ssh

Author: I also introduced a "finish line" to make it clear that from there downwards is only to 'enhance knowledge', not to 'get it started'

`Vagrant` is used as a tool to define VMs characteristics, that will be then run through Virtualbox - in other words, it is used so we can write a manifesto that defines the characteristics of a VM, and the VM generated has always the same characteristics. It is useful in this case to model the development VM just as if it was the server on which we will run the application.

> `Make` is a tool that, among other things, chains commands together. So, for instance, you write in the `Makefile` that `a` runs a specific long command, `b` a different long command, and you can call the commands with `make a` or `make b`. You can also write a command `c` which is a chain of `a` followed by `b`. We use it to set a 'flow' of operations that should be followed (e.g. as explained, `make bootstrap` chains 3 operations, and one such operation is used very often i.e. `build` and/or `start`). This is used on the _GUEST_.

Reviewer: ```suggestion
`Make` is a tool that, among other things, chains commands together, automatically handling dependencies between those commands. So, for instance, you write in the `Makefile` that `a` runs a specific long command, `b` a different long command, and you can call the commands with `make a` or `make b`. You can also write a command `c` which is a chain of `a` followed by `b`. We use it to set a 'flow' of operations that should be followed (e.g. as explained, `make bootstrap` chains 3 operations, and one such operation is used very often i.e. `build` and/or `start`). This is used on the _GUEST_.
```
Is the makefile indented to be only ever executed on the GUEST? Then I'd like a command stating so on the top of the `Makefile` file. "Those commands are intended to be executed on the GUEST system, i.e. within the vagrant VM".

I think `make bootstrap` is intended to be executed on the HOST?

Author: no, `make bootstrap` is executed *by* vagrant on the GUEST (so, one never even notices)

Added your suggestion, and in the README is also explained at the section where it says what is the `Makefile`

Author: no to your second question, yes to your first question

`start.sh` runs either `vagrant up` or `make bootstrap` (according to how you want to run your system in local) so one has to literally only launch one command and it's set to be working, after the startup time of around 10-20 minutes (according to internet connection speed). This is used on the _HOST_.

### `start.sh` and Makefile

> `start.sh`

On the _HOST_, i.e. the machine that runs the virtual machine, you use `start.sh` which can either:
- Start the vm
  - Use `./start.sh` for normal development cycle: app runs in development mode
  - Use `./start.sh --fast` for sysops/integration development cycle: app runs in production mode so you can concentrate on developing integration to the app, not the app itself

> - Reset the settings to recreate the virtual machine (`./start.sh --reset`). This is in case you experimented so hard that you made something exploooode

Reviewer: ```suggestion
- Reset the settings to recreate the virtual machine (`./start.sh --reset`). This is in case you experimented so hard that you made something exploooode. Doing so, you will lose the users and other content you created on your local instance of MyAegee, but this will not remove your source code.
```

Author: Nice, added

> ### docker-compose.yml

Reviewer: comment for line 204: I would add the steps for the example 1 and 2. Maybe not in the main README, but somewhere. Simply state to use `docker stop myaegee_mailer_1`

For example 2: You added a comment in the `.env` file to edit the array. I guess the next step is to run a `make start` on the GUEST, right?

considering how small the answers are, I'd add them here actually.

In the docker-compose files there are the definitions of where an app should be reached.

Docker-compose will use the variables defined above, and put them under the `labels` section of a container (if a container needs it). The `labels` section is parsed by traefik to route all the HTTP calls to the correct containers. In other words, this is where the values contained in the `.env` file are used to specify that the app replies on '`my.`appserver.test' instead of e.g. '`magic.`appserver.test'.

### /etc/hosts

This file which is located on the _HOST_ machine is used to add new [subdomains for the services](#subdomains-registered-on-traefik): e.g. add `pgadmin` to be able to go to `pgadmin.appserver.test` and use a visual tool for the databases. The important part is that this name of subdomain matches the one defined in the file `.env`

# Troubleshooting / other

> - If you are stuck and something doesn't seem to work, make sure you don't have a mismatch between `/etc/hosts`, the URL you type in the browser, and the address that the system expects. See file `current-config.yml` for that: it is a file which is generated every time `make start` is launched. It contains the description of the desired state of the app.

Reviewer: this file is visible on the HOST and GUEST since the filesystem is mounted, right?

Author: `/etc/hosts` is the only HOST-only file

Author: i added it

Reviewer: sorry, I meant the `current-config.yml`, not the /etc/hosts. I didn't realize there were two files in this paragraph ;D

> 

Reviewer: I'd add here the `vagrant ssh`, `docker ps` as a troubleshooting tool. It can help understanding a /etc/hosts file misconfig.

- You can add dev-tools to the array of services and troubleshoot docker using portainer (mentioned above). Make sure that you have added also the URL to the hostfile (`/etc/hosts`) and that it matches the variable defined in `.env`

> - If you remove services from array `ENABLED_SERVICES`, they are not stopped. This is not a problem in general, just don't be surprised if when you start (having the configuration `ENABLED_SERVICES=core:frontend:events`) and stop later (with the configuration `ENABLED_SERVICES=events`) you find a message `WARNING: found orphan containers`

Reviewer: ```suggestion
- If you remove services from array `ENABLED_SERVICES`, they are not stopped. This is not a problem in general, just don't be surprised if when you start (having the configuration `ENABLED_SERVICES=core:frontend:events`) and stop later (with the configuration `ENABLED_SERVICES=events`) vagrant writes a message `WARNING: found orphan containers`.
```

Author: not vagrant but docker does, I added the rephrasing anyhow

## How to run unit/integration tests when you start the system

> ?? @serge1peshcoff

Author: @serge1peshcoff a question for you

Reviewer 3: Short answer on how to run tests on host for all services except mailer:

(need to do one time)
- install node
- npm install in the app folder
- install postgres (no idea how to do it on win and Linux, on Mac I use Postgres.app. probably docker container will do, check out how I do it in .circleci/config.yml for CI)

Run tests:
- npm test

Run lint:
- npm run lint

Should be enough. @linuxbandit

Reviewer: for postgres on Linux, I did it once. I'll look up my history; I did use docker for that, it worked quite well. I'll figure out which config I had to change locally.

Author: perhaps we can create a target of `helper.sh` that could e.g. run `--test` ?

Author: well it'd suck

> ?? @serge1peshcoff

Reviewer 3: I'll write this later

Author: ok, there are two questions for you

## Default credentials

From the [core readme](./core/README.md): there are test users. All users have `5ecr3t5ecr3t` as password.

- `admin@example.com` - admin
- `board@example.com` - board member of antenna
- `member@example.com` - regular member of antenna
- `not-confirmed@example.com` - member who is not confirmed
- `password-reset@example.com` - member who requested a password reset
- `suspended@example.com` - a suspended member

You can use `5ecr3t` for a password reset token (for a member with email `password-reset@example.com`) and `5ecr3t` for a mail confirmation (for a member with email `not-confirmed@example.com`).

**NOTICE** if you use the `--fast` mode of `start.sh`, then core does NOT provision in production mode, so you will be fast in bootstrapping but without users to play with.

> @serge1peshcoff please confirm

Author: @serge1peshcoff a question for you

Reviewer 3: Yes

Author: Also the admin is not provisioned at all?
Can I tell people to register normally if they need a user account, or would they be blocked by an email?
Is there a "quick" way for them to make the migration run even in production mode, or to provision any user without registration?

## How to reset the database and recreate

A couple of options

1) As mentioned above, you can enable dev-tools for pgadmin. From there you can delete the db
  - Notice unfortunately you have to configure pgadmin when you first login: specify the host, username, password.
    - Username and password of the database is NOT the same username and password of pgadmin. Find everything in the files mentioned in 'moving parts', or `current-config.yml` for ease.
    - You put the name of the service as hostname (docker internally resolves stuff with its internal DNS). In other words, every container is reachable at the host named like its service (e.g. `frontend`, `core`, `postgres-core`, etc). The 'service' is named in the `docker-compose.yml` under the key `services:`. Again: _the title of the container is also the hostname of the container_

2) Use portainer to delete the service and db, then `make start` to fix everything.

>   - The service must be deleted/restarted because it runs the migrations and therefore fills the DB with the important data. If you delete only the db, the service will expect the db to be filled with data, causing errors

Reviewer: I'll enhance this with a few words on how to run a manual `SELECT * from table` with `pgadmin`. I'll add an example for the steps we did today together too.

Reviewer 3: do we have pgadmin working though?
